### ISSUE TITLE
[Python] `Tracecontext` OTel propagator config generates an additional X-Ray span

### ISSUE THREAD

**Issue by michaelhyatt** [aws-observability/aws-otel-lambda#123]
I am using v1.3.0 of the Python layer and configured the lambda to use `tracecontext` propagation, but I can see the lambda reporting an additional extra span that seems to be related to X-Ray, which shouldn't be there.

Notable additional config:
```
    OTEL_PROPAGATORS: "tracecontext"
    OTEL_TRACES_SAMPLER: "Always_On"
```

Lambda tracing is turned off (PassThrough).

Code:
```python
def consumer(event, lambda_context):

    context = extract(event['headers'])

    # Create top level transaction representing the lambda work
    with tracer.start_as_current_span("consumer-function-top-level", context=context, kind=SpanKind.SERVER):

        # Some other internal work performed by the lambda
        with tracer.start_as_current_span("consumer-function-some-internal-work", kind=SpanKind.INTERNAL):

            time.sleep(1)

        time.sleep(0.5)

    return {'statusCode': 200}
```

This should produce 2 spans: top-level `consumer-function-top-level` and nested `consumer-function-some-internal-work`. Instead, I am seeing 3 spans created in CloudWatch log with Span 2 `065c07aa6cc4916d` that shouldn't be there:
```log


2021-07-24T04:36:51.316Z	INFO	loggingexporter/logging_exporter.go:41	TracesExporter	{     "#spans": 3 }
--
2021-07-24T04:36:51.316Z	DEBUG	loggingexporter/logging_exporter.go:51	ResourceSpans #0
Resource labels:
-> telemetry.sdk.language: STRING(python)
-> telemetry.sdk.name: STRING(opentelemetry)
-> telemetry.sdk.version: STRING(1.3.0)
-> cloud.region: STRING(ap-southeast-2)
-> cloud.provider: STRING(aws)
-> faas.name: STRING(aws-python-api-worker-project-dev-consumer)
-> faas.version: STRING($LATEST)
-> service.name: STRING(aws-python-api-worker-project-dev-consumer)
InstrumentationLibrarySpans #0
InstrumentationLibrary handler

Span #0
Trace ID       : ac4d7110c6d22e985b5db28e8ecca769
Parent ID      : 3c374013084a679c
ID             : 522890b38766e922
Name           : consumer-function-some-internal-work
Kind           : SPAN_KIND_INTERNAL
Start time     : 2021-07-24 04:36:49.813522476 +0000 UTC
End time       : 2021-07-24 04:36:50.814894967 +0000 UTC
Status code    : STATUS_CODE_UNSET
Status message :

Span #1
Trace ID       : ac4d7110c6d22e985b5db28e8ecca769
Parent ID      : e24504245b7cff0e
ID             : 3c374013084a679c
Name           : consumer-function-top-level
Kind           : SPAN_KIND_SERVER
Start time     : 2021-07-24 04:36:48.812135568 +0000 UTC
End time       : 2021-07-24 04:36:51.315562096 +0000 UTC
Status code    : STATUS_CODE_UNSET
Status message :

Span #2 <==== EXTRA SPAN!!!
Trace ID       : ac4d7110c6d22e985b5db28e8ecca769
Parent ID      : ad1bb604de1da312
ID             : 065c07aa6cc4916d
Name           : handler.consumer
Kind           : SPAN_KIND_SERVER
Start time     : 2021-07-24 04:36:48.811673087 +0000 UTC
End time       : 2021-07-24 04:36:51.315622948 +0000 UTC
Status code    : STATUS_CODE_UNSET
Status message :
Attributes:
-> faas.execution: STRING(e759abdb-6119-4f91-afe0-716a71276825)
-> faas.id: STRING(arn:aws:lambda:ap-southeast-2:401722391821:function:aws-python-api-worker-project-dev-consumer)
-> faas.name: STRING(aws-python-api-worker-project-dev-consumer)
-> faas.version: STRING($LATEST)


```

**Comment by NathanielRN** [comment]
Hey @michaelhyatt! Thanks for providing logs and steps to reproduce.

I don't think that extra span has anything to do with X-Ray, but if you think it does can you let me know?

I also think that what you are seeing is the expected behavior, if you are following the public AWS OTel docs on [Getting Started with Lambda for Python](https://aws-otel.github.io/docs/getting-started/lambda/lambda-python).

This is because the docs include this step:

> Add the environment variable AWS_LAMBDA_EXEC_WRAPPER and set to /opt/otel-instrument.

This environment variable is used to run the `/opt/otel-instrument` script _before_ your `def consumer(event, lambda_context)` function. That script in `opentelemetry-instrument` will in turn [call a wrapper otel-instrument Python script](https://github.com/open-telemetry/opentelemetry-lambda/blob/3f5d7b8bcbc5d2b9cb6283fdb7c772414045e90b/python/src/otel/otel_sdk/otel-instrument#L13) which will **automatically instrument Lambda by creating a trace.**

Given the `faas` namespace in the attributes of Span#2, and its timestamp which both _starts_ and _ends_  before/after Spans#0 and Spans#1 I suspect that this is the span created [by the otel_wrapper.py file](https://github.com/open-telemetry/opentelemetry-lambda/blob/3f5d7b8bcbc5d2b9cb6283fdb7c772414045e90b/python/src/otel/otel_sdk/opentelemetry/instrumentation/aws_lambda/__init__.py#L111-L133) to encompass them as the overarching parent span.

I was able to see _only_ your 2 spans (I used the SDK exporter not the Collector exporter though) by not setting the `AWS_LAMBDA_EXEC_WRAPPER` environment variable and configuring the `opentelmetry` package on Lambda on my own without the help of the `otel-instrument` script. From what I understand, that script is meant to initialize much of `opentelmetry` for you automatically and give you a top "Lambda" span (Span#2) from which any subsequent spans will follow so you can see where they descend from on the Service Map. But if you have another use case, the layer is still useful because it provides all the `opentelemetry` package code for you to be able to configure it how you want 🙂 

I hope that provides a solution for you, please let me know if I can answer anything else!

**Comment by michaelhyatt** [comment]
Hi @NathanielRN, thanks for looking into it.

I can see that the span is indeed created by the `otel_wrapper.py` file. It automatically assumes X-Ray trace context propagation [here](https://github.com/open-telemetry/opentelemetry-lambda/blob/3f5d7b8bcbc5d2b9cb6283fdb7c772414045e90b/python/src/otel/otel_sdk/opentelemetry/instrumentation/aws_lambda/__init__.py#L106-L109) and I can see that the generated span (no 2 in my log) has `parent.id` taken from the X-Ray trace ID header.

In my case, I am using a different trace context propagation method. Instead of X-Ray, I am relying on W3C `traceparent` header, hence is the setting `OTEL_PROPAGATORS: "tracecontext"`. The problem is [this code](https://github.com/open-telemetry/opentelemetry-lambda/blob/3f5d7b8bcbc5d2b9cb6283fdb7c772414045e90b/python/src/otel/otel_sdk/opentelemetry/instrumentation/aws_lambda/__init__.py#L106-L109) is hardwired to propagate the X-Ray trace ID instead of the one I am sending in the `traceparent` HTTP header.

In my previous example you can see the mismatch between the parent.ids of the spans:
* Span1.parent.id is injected from `traceparent` header (not shown there).
* Span0.parent.id is Span1.ID <= this is done by my instrumentation.
* Span2.parent.id is pointing to some other external span.id that is presumably coming from some X-Ray external trace.

So, to summarise, I am perfectly happy with the auto-instrumentation of my lambda code, as long as the trace context is propagated properly respecting the `OTEL_PROPAGATORS` setting instead of being hard-wired to AWS X-Ray.

**Comment by NathanielRN** [comment]
@michaelhyatt sure thing! I spoke with my colleague @garrettwegan and I hope I can give you an answer 🙂 

Thanks for pointing me to that code! You're absolutely right that there is an assumption that the `_X_AMZN_TRACE_ID` environment variable contains the context. (Instead of looking at HTTP headers).

I should point out that the code you linked to is in an _upstream OpenTelemetry repository_ and not in this downstream aws-otel-lambda repository (although this repo does use that code from the upstream). The code you see there was contributed as part of our team's initial release to get OpenTelemetry working with Python (and other SDKs) and the initial scope included making sure it worked with AWS X-Ray, which is why you see that it is hard-coded.

Some OTel Lambda implementations in other languages have already evolved to allow for a wider amount of user use-cases like yours. For example, the upstream OTel Lambda JavaScript now gives users the [option to disables AWS X-Ray Propagation and use custom code to extract the context out of the Lambda event](https://github.com/open-telemetry/opentelemetry-js-contrib/tree/68e54490b207980b080d6642be48e81d17659cc7/plugins/node/opentelemetry-instrumentation-aws-lambda#aws-lambda-instrumentation-options). Then you can use that custom context form the Lambda event instead of having it look for an X-Ray context from the `_X_AMZN_TRACE_ID` env variable by default.

For your issue, my suggestion would be to **open an issue upstream** to update the OTel Python instrumentation to have an option similar to JavaScript. The repo is still in early stages of development so it is taking help to make all the languages better!

As another note, you said this:

> instead of the one I am sending in the traceparent HTTP header.

I think you may run into problems here. I _have not verified this myself_, but from what I understand, the design of Lambda is that it will _never_ have HTTP headers because _Lambda cannot be directly invoked by an HTTP request._ HTTP headers are replaced by `_X_AMZN_TRACE_ID` for X-Ray (as a special case) or by "Lambda Events" (in any other case). So you **must** get your context from a Lambda Event if you have a custom context you want to use. (i.e. you would invoke your Lambda function and insert the HTTP headers into the event that is to be received at the Lambda side).

The ideal future would be one where upstream supports both the X-Ray method of using `_X_AMZN_TRACE_ID` and the poweruser method way like yourself that uses Lambda Events and an "`eventContextExtractor`" (like OTel JS) to get the context.

Please let me know if that explains the issue! Unfortunately it's just a limitation of where the project is for now 😞

**Comment by michaelhyatt** [comment]
Thanks, @NathanielRN and @garrettwegan. I created [this issue](https://github.com/open-telemetry/opentelemetry-lambda/issues/123) in the upstream `opentelemetry-lambda` repo.

I wanted to clarify the trace context propagation with Lambdas.

1. Lambda function receives HTTP headers through `event["headers"]` and using OTel tracecontext `extract`/`inject` with the right context propagator setting should take care of using the right getter and setter to extract the parent trace context. A very similar setup can work with SQS message headers and any other protocol supporting metadata attached to the event. So, AWS X-Ray context propagation shouldn't be the only way to propagate trace context between services.
2. I had a similar [issue](https://github.com/aws-observability/aws-otel-lambda/issues/118) with JS lambdas, but now it has been fixed with v0.23 of JS layer and `tracecontext` propagation appears to work as expected.
3. I really like all the initialisation that is done within `otel-instrument` and would like to continue to rely on it without having to initialise OTel myself. And relying on auto instrumentation would be super helpful instead of creating the spans myself, however, I really want the spans to be linked to the external context rather than AWS X-Ray.

Hope it makes sense, happy to move the discussion into the [other upstream issue](https://github.com/open-telemetry/opentelemetry-lambda/issues/123).

**Comment by NathanielRN** [comment]
Thanks for creating that issue! I appreciate it! 😄 

I'll do my best to answer your questions:

Regarding 1, I don't think you can count on headers being in `event["headers"]`. It seems like you can configure that for API Gateway but it's not defined? It would help to continue [this thread on my PR to fix the problem](https://github.com/open-telemetry/opentelemetry-lambda/pull/124#discussion_r679426011) if you know otherwise!

Regarding 2, I'm glad you got a solution! However I looked at the issue you posted and I'm not sure what fix was done to address that? But it's probably best to continue the conversation there if you have any additional questions.

Regarding 3, I think we may encounter difficulties again as pointed out in [the thread on my PR](https://github.com/open-telemetry/opentelemetry-lambda/pull/124#discussion_r679426011) but hopefully we do find  a solution upstream that works!
